**Content list:** 

1. Importing libraries
2. Searching for mixed data type, missing values and duplicates in the products dataset and fixxing them.  
3. Searching for mixed data type, missing values and duplicates in the orders dataset and fixxing them.  

In [ ]:
import numpy as np
import pandas as pd

In [ ]:

df_prods=pd.read_csv('products.csv')

In [ ]:
df_ords=pd.read_csv('orders_wrangled.csv',index_col=False)

In [ ]:
#Checking for mixed-type data in df_prods dataframe

for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


In [ ]:
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [ ]:
#Checking for missing values in df_prods dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

There are 16 missing values in the product_name column.

In [ ]:
df_prods.shape

(49693, 5)

In [ ]:
#Creating a separate dataframe for the missing values
df_prod_nan=df_prods[df_prods['product_name'].isnull() == True]

In [ ]:
df_prod_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [ ]:
df_prod_nan.shape

(16, 5)

In [ ]:
#Creating a new dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [ ]:
df_prods_clean.shape

(49677, 5)

In [ ]:
#Checking for duplicates in df_prods dataframe

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [ ]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [ ]:
df_dups.shape

(5, 5)

In [ ]:
#Creating a new dataframe without duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [ ]:
df_prods_clean_no_dups.shape

(49672, 5)

In [ ]:
df_ords.describe() 

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

The unnamed column is the index column assigned by pandas

In [ ]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


In [ ]:
df_ords.shape

(3421083, 7)

In [ ]:
#Checking for mixed-type data in df_ords dataframe

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There are apparently no mixed columns in the df_ords dataframe.

In [ ]:
#Checking for missing values in df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the days_since_prior_order column. Perhaps, the dates of orders placed by each customer weren't recorded in the system properly. As a result, the differences of days between orders couldn't be calculated for all entries. 

In [ ]:
from numpy.core.fromnumeric import mean
mean(df_ords['days_since_prior_order'])

11.114836226863012

In [ ]:
df_ords['days_since_prior_order'].median()

7.0

In [ ]:
df_ords['days_since_prior_order']

0           NaN
1          15.0
2          21.0
3          29.0
4          28.0
           ... 
3421078    29.0
3421079    30.0
3421080    18.0
3421081     7.0
3421082    30.0
Name: days_since_prior_order, Length: 3421083, dtype: float64

In [ ]:
#Filling the missing values with median
df_ords['days_since_prior_order'].fillna(df_ords['days_since_prior_order'].median(),inplace=True)

The missing values in the days_since_prior_order column have been replaced with the median of the column, since the mean would likly be biased by extreme values.

In [ ]:
#Checking for missing in df_ords

df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

All the missing values have been succesfully replaced. 

In [ ]:
#Checking for duplicates in df_ords dataframe

df_ords_dups = df_ords[df_ords.duplicated()]

In [ ]:
df_ords_dups.shape

(0, 7)

In [ ]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

There are apparently no duplicates in the df_ords dataframe.

In [ ]:
#Creating a new dataframe without duplicates

df_ords_clean_no_dups = df_ords.drop_duplicates()

In [ ]:
df_ords_clean_no_dups.shape

(3421083, 7)

After using the command to remove duplicates, the rows and column number are same in the df_ords dataframe. Which means there were no duplicates in this dataframe.

In [ ]:
df_ords_clean_no_dups

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
1                 1   2398795        1             2          3   
2                 2    473747        1             3          3   
3                 3   2254736        1             4          4   
4                 4    431534        1             5          4   
...             ...       ...      ...           ...        ...   
3421078     3421078   2266710   206209            10          5   
3421079     3421079   1854736   206209            11          4   
3421080     3421080    626363   206209            12          1   
3421081     3421081   2977660   206209            13          1   
3421082     3421082    272231   206209            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [ ]:
from google.colab import files
df_prods_clean_no_dups.to_csv('df_prods_clean.csv') 
files.download('df_prods_clean.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:

df_ords_clean_no_dups.to_csv('df_ords_clean.csv') 
files.download('df_ords_clean.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>